design/0005-circuit-breaking.md

```diff
@@ -0,0 +1,231 @@
+# Title
+
+- Contribution Name: circuit-breaking
+- Implementation Owner: @Ashish-Bansal, Linkerd Team
+- Start Date: 2020-06-14
+- Target Date: N.A.
+- Linkerd Issue:
+  [linkerd/linkerd2#2846](https://github.com/linkerd/linkerd2/issues/2846)
+- Reviewers: @adleong
+
+## Summary
+
+[summary]: #summary
+
+We would like to implement circuit breaking functionality which would allow end
+users to configure mesh such that they can limit the impact of failures from
+faulty service over others and prevent cascading issues. By configuring circuit
+breaking, they will be able to remove failing service instances from load
+balancing and survive through it in a more graceful manner.
+
+## Problem Statement
+
+In a service mesh, services often make calls to other services. In case upstream
+service is unable to respond due to any reason, it's possible that these
+failures can cascade to other services in the cluster.
+
+If we are aware that specific service instance is not responding, we can call it
+unhealthy and it's better if we just stop sending "traffic"[0] to that instance
+rather than getting just another expected failure.
+
+---
+
+It would also be somewhat helpful in case of canary deployments when there's
+some regression and new build is raising failures.
+
+[Example](https://drive.google.com/open?id=1FxKi-egyhHumCvlHvTI4P114z7qOLPeN)
+
+---
+
+Users should also be able to set max-ejection-percentage which provides max
+number of service instances which can be removed from the load balancing for
+specific service.
+
+[Example](https://drive.google.com/open?id=1jSMaogdvzxjv0v2dBoL7OYGHmsCusXn-)
+
+---
+
+Network Parition - There can be lot of scenarios in network partition but to
```

Comment:
> not have anything to do with connection errors

In case of network errors i.e. we are not able to reach the host itself, yes it won't involve circuit breaking. It's similar to how we are handling things in case of TCP. 

In case we do plan to handle circuit breaking for network errors or at any other layer, it will be handled differently. Here we will just be handling issues arising from the HTTP layer (or gRPC etc). Anything happening below application layer, nope, we won't be handling it.